### PR TITLE
✨ Add filters to HfApi.get_repo_discussions

### PR DESCRIPTION
**Pull request by SBrandeis**
# TL;DR

Add new filters to the `HfApi.get_repo_discussions` method

This PR depends on https://github.com/huggingface/moon-landing/pull/8147 (private) for the `author` filter.

Ping @Narsil 

## How it works

```python
import huggingface_hub

discussion_generator = huggingface_hub.get_repo_discussions(
    repo_id="openai/whisper-large-v3",
    author="sanchit-gandhi",
    discussion_type="pull_request",
    discussion_status="open",
)

for discussion in discussion_generator:
    print(discussion)
    # ^ Only open pull requests authored by `sanchit-gandhi`
```

### CODE REVIEW

File: src/huggingface_hub/hf_api.py [hunk 1]
```diff
@@ -5191,6 +5193,9 @@ def get_repo_discussions(
         self,
         repo_id: str,
         *,
+        author: Optional[str] = None,
+        discussion_type: Optional[str] = None,
+        discussion_status: Optional[str] = None,
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:5198]: I'd add type annotation here to help autocompletion tools

```suggestion
        discussion_type: Optional[Literal[["all", "discussion", "pull_request"]] = None,
        discussion_status: Optional[Literal["all", "open", "closed"]] = None,
```

**Review by Wauplin — COMMENTED** [review]
Thanks for the direct PR @SBrandeis :heart: 

Looks good to me. 2 things I would add:
- mention in the docs that it's possible to filter by type, status and author. A small paragraph + a unique example with all 3 filters at once is enough (somewhere [here](https://github.com/huggingface/huggingface_hub/blob/f42b21a183e07ad97e96edc3a02f522a0b13af79/docs/source/en/guides/community.md?plain=1#L61)).
- add tests to check filter are correctly sent to the server. Something like the snippet below [here](https://github.com/huggingface/huggingface_hub/blob/f42b21a183e07ad97e96edc3a02f522a0b13af79/tests/test_hf_api.py#L2423).

```py
def test_get_repo_discussion_by_type(self):
    discussions = list(self._api.get_repo_discussions(repo_id=self.repo_id, discussion_type="discussion"))
    self.assertEqual(len(discussions), 1)
    self.assertEqual(discussions[0].num, self.discussion.num)

def test_get_repo_discussion_by_author(self):
    discussions = list(self._api.get_repo_discussions(repo_id=self.repo_id, discussion_type=USER))
    self.assertEqual(len(discussions), 2)
    
    discussions = list(self._api.get_repo_discussions(repo_id=self.repo_id, discussion_type="julien-c"))
    self.assertEqual(len(discussions), 0)
```

**Review by julien-c — APPROVED** [review]
looks good!

File: docs/source/en/guides/community.md [hunk 2]
```diff
@@ -25,6 +25,21 @@ The `HfApi` class allows you to retrieve Discussions and Pull Requests on a give
 [...]
 ```
 
+`HfApi.get_repo_discussions` supports by author, type (Pull Request or Discussion) and status (`open` or `closed`):
```
Comment (Wauplin) [inline docs/source/en/guides/community.md:28]:
```suggestion
`HfApi.get_repo_discussions` supports filtering by author, type (Pull Request or Discussion) and status (`open` or `closed`):
```

File: src/huggingface_hub/constants.py [hunk 3]
```diff
@@ -79,6 +80,10 @@ def _as_int(value: Optional[str]) -> Optional[int]:
     "models": REPO_TYPE_MODEL,
 }
 
+DiscussionTypeFilter = Literal["all", "discussion", "pull_request"]
+DISCUSSION_TYPES: Tuple[DiscussionTypeFilter, ...] = typing.get_args(DiscussionTypeFilter)
+DiscussionStatusFilter = Literal["all", "open", "closed"]
+DISCUSSION_STATUS: Tuple[DiscussionTypeFilter, ...] = typing.get_args(DiscussionStatusFilter)
```
Comment (Wauplin) [inline src/huggingface_hub/constants.py:86]: :exploding_head:

File: tests/test_hf_api.py [hunk 4]
```diff
@@ -2427,6 +2427,52 @@ def test_get_repo_discussion(self):
...
+    def test_get_repo_discussion_by_author(self):
+        discussions_generator = self._api.get_repo_discussions(repo_id=self.repo_id, author="no-discussion")
+        self.assertIsInstance(discussions_generator, types.GeneratorType)
+        self.assertListEqual(list([d.num for d in discussions_generator]), [])
+
+        discussions_generator = self._api.get_repo_discussions(repo_id=self.repo_id, author=USER)
+        self.assertIsInstance(discussions_generator, types.GeneratorType)
+        self.assertListEqual(
+            list([d.num for d in discussions_generator]), [self.discussion.num, self.pull_request.num]
+        )
```
Comment (Wauplin) [inline tests/test_hf_api.py:2457]:
```suggestion
    def test_get_repo_discussion_by_author(self):
        discussions_generator = self._api.get_repo_discussions(repo_id=self.repo_id, author="unknown")
        self.assertIsInstance(discussions_generator, types.GeneratorType)
        self.assertListEqual(list([d.num for d in discussions_generator]), [])
```

Let's not add a failing test. This should be fine

**Review by Wauplin — APPROVED** [review]
Thanks a lot for the changes @SBrandeis! Looks good to me :)

**Pull request merged**